### PR TITLE
Orthogonal defaulting and new pruning algorithm

### PR DESCRIPTION
**Pull request by sttts**
This PR allows to implement different algorithms orthogonally to the main
validation recursion. The Result object tracks the validating schemata for
each node in the input JSON. This means that these algorithms do not
have to care about quantors or the different ways to specify properties.

This means:

- defaulting becomes a trival 12 liner:
```golang
// ApplyDefaults applies defaults to data.
func ApplyDefaults(r *validate.Result) {
	fieldSchemata := r.FieldSchemata()
	for key, schemata := range fieldSchemata {
	LookForDefaultingScheme:
		for _, s := range schemata {
			if s.Default != nil {
				key.Object()[key.Field()] = s.Default
				break LookForDefaultingScheme
			}
		}
	}
}
```
- pruning of unspecified fields is implementable in a similar way. It's added as another example.

Running defaulting benchmarks 30000 leads to:
```
Before PR:
30000	     44573 ns/op	   54581 B/op	     499 allocs/op
30000	     43208 ns/op	   54581 B/op	     499 allocs/op
30000	     39850 ns/op	   54582 B/op	     499 allocs/op
30000	     39427 ns/op	   54582 B/op	     499 allocs/op
```
After PR:
```
30000	     48023 ns/op	   66088 B/op	     497 allocs/op
30000	     46331 ns/op	   66086 B/op	     497 allocs/op
```

In other words: runtime is 15% higher, memory is 20% higher. If this is a problem, we could add a second `ValidateWithSchemaTracking()` func which does the extra processing and returns a result that has the necessary getters.

TODO:
- [x] benchmark cpu and memory/allocations:

### CODE REVIEW

**Comment by fredbi** [comment]
I like the idea of metadata in the result structure. 
That is something I find useful to spew more accurate messages on _where_ the validation error did occur.
But I am not sure this is what you had in mind.
Your PR is lagging master for several months I think.

**Comment by sttts** [comment]
> Your PR is lagging master for several months I think.

Rebased already.

Will push the second algorithm soon to show what this is all about.

**Comment by sttts** [comment]
@fredbi compare the last commit. This is the kind of trivial algorithm I have in mind, without complicating the main validation recursion.

File: defaulter.go [hunk 1]
```diff
@@ -14,16 +14,40 @@
...
+		for _, val := range obj {
+			applyDefaults(val, result)
+		}
+	}
 }
 
-// DefaulterFunc is a function to be called to apply default values to an object.
-type DefaulterFunc func()
+func applyObjectDefaults(obj map[string]interface{}, result *Result) {
+	for fld, schemata := range result.propertySchemata[fmt.Sprintf("%p", obj)] {
```
Comment (sttts) [inline defaulter.go]: @s-urbaniak this is what I talked about in.

**Comment by sttts** [comment]
@fredbi please take another look. This is ready for review.

**Comment by casualjim** [comment]
this is causing issues downstream, I'm going to revert this PR.. Will try to add the failing cases to this CI build so we discover this earlier.

**Comment by nikhita** [comment]
> this is causing issues downstream

@casualjim Can you elaborate on this?

**Comment by casualjim** [comment]
After applying this PR, the types in slice properties are no longer []interface{}

Here's a job that fails:
https://travis-ci.org/go-openapi/runtime/jobs/366714797


just running the tests in go-openapi/runtime shows the issues.

**Comment by sttts** [comment]
@casualjim will look into that tomorrow.

Some integration CI would be nice. How do you notice those problems now? Just merging and then seeing regression on the next downstream runs?

**Comment by casualjim** [comment]
yeah basically.
I would like to add the following to .travis.yml

```yaml
- go get -u github.com/go-openapi/runtime
- cd $GOPATH/src/github.com/go-openapi/runtime
- go test ./...
```

to run after the tests of the validate repo have finished

**Comment by sttts** [comment]
I took a quick look at the issue: my change in the `schemaSliceValidator` adds the requirement that the input `data` is a JSON like `[]interface{}` slice, but here it is not: it's a `[]middleware.friend`. We used reflection before my change to work around this. 

I noticed that the `objectValidator` in contrast has a hard requirement on  `map[string]interface{}` which sounds correct.

Am I right that the `data` used as validator input must be JSON? I.e. not an arbitrary Go type? If yes, the runtime middleware test broken it seems.

**Comment by casualjim** [comment]
yes you're right, the input should be json for the json schema validators

**Comment by sttts** [comment]
Working on the go-openapi/runtime fix now.
